Feat/add extract primitive generator

elite-benni opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our
      guidelines: https://github.com/goetzrobin/spartan/blob/main/CONTRIBUTING.md#-commit-message-guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [x] Documentation content changes
- [ ] Other... Please describe:

## Which package are you modifying?

- [x] accordion
- [x] alert
- [x] alert-dialog
- [x] aspect-ratio
- [x] avatar
- [x] badge
- [x] breadcrumb
- [x] button
- [x] calendar
- [x] card
- [x] carousel
- [x] checkbox
- [x] collapsible
- [x] combobox
- [x] command
- [x] context-menu
- [x] data-table
- [x] date-picker
- [x] dialog
- [x] dropdown-menu
- [x] form-field
- [x] hover-card
- [x] icon
- [x] input
- [x] input-otp
- [x] label
- [x] menubar
- [x] navigation-menu
- [x] pagination
- [x] popover
- [x] progress
- [x] radio-group
- [x] scroll-area
- [x] select
- [x] separator
- [x] sheet
- [x] skeleton
- [x] slider
- [x] sonner
- [x] spinner
- [x] switch
- [x] table
- [x] tabs
- [x] textarea
- [x] toggle
- [x] toggle-group
- [x] tooltip
- [x] typography

## What is the current behavior?

sometimes the examples won't work because the defaultCode example was not updated correctly. 

Closes #735 

## What is the new behavior?

the code Snippets can be generated by an nx generator, which takes away the pain of keeping in up to date. 
The generation happens on serve and build and of course can be triggered manually.
Added a script in package.json ```generate-snippets```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

elite-benni commented:
Sorry for the size of the PR

goetzrobin commented on libs/tools/src/generators/generate-primitive-snippets/generator.spec.ts:
Did the same for the test. Less confident if this is useful here:

```ts
import { joinPathFragments, Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { extractPrimitiveCodeGenerator } from './generator';

describe('extractPrimitiveCodeGenerator', () => {
  let tree: Tree;
  const componentsBasePath = 'apps/app/src/app/pages/(components)/components';

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  function createMockPreviewFile(primitiveName: string, content: string): string {
    const filePath = joinPathFragments(componentsBasePath, primitiveName, `${primitiveName}.preview.ts`);
    tree.write(filePath, content);
    return filePath;
  }

  function createMockExampleFile(primitiveName: string, exampleName: string, content: string): string {
    const filePath = joinPathFragments(componentsBasePath, primitiveName, `${exampleName}.example.ts`);
    tree.write(filePath, content);
    return filePath;
  }

  function getGeneratedFileContent(primitiveName: string): string | null {
    const filePath = joinPathFragments(componentsBasePath, primitiveName, `${primitiveName}.generated.ts`);
    return tree.read(filePath, 'utf-8');
  }

  function expectGeneratedFileExists(primitiveName: string): void {
    const filePath = joinPathFragments(componentsBasePath, primitiveName, `${primitiveName}.generated.ts`);
    expect(tree.exists(filePath)).toBe(true);
  }

  function expectGeneratedFileNotExists(primitiveName: string): void {
    const filePath = joinPathFragments(componentsBasePath, primitiveName, `${primitiveName}.generated.ts`);
    expect(tree.exists(filePath)).toBe(false);
  }

  const mockAngularComponent = `
import { Component } from '@angular/core';
import { NgIcon, provideIcons } from '@ng-icons/core';
import { lucideBox } from '@ng-icons/lucide';
import { HlmAlertDirective } from '@spartan-ng/helm/alert';

@Component({
  selector: 'spartan-alert-preview',
  imports: [HlmAlertDirective, NgIcon],
  providers: [provideIcons({ lucideBox })],
  template: \`
    <div hlmAlert>
      <ng-icon name="lucideBox" />
      <h4>Test Alert</h4>
    </div>
  \`
})
export class AlertPreviewComponent {}
`;

  const mockComponentWithUnwantedExports = `
import { Component } from '@angular/core';

@Component({
  selector: 'test-component',
  template: '<div>Hello World</div>'
})
export class TestComponent {}

// These should be removed by the generator
export const defaultSkeleton = \`<div>skeleton</div>\`;
export const defaultImports = \`import stuff\`;
export const someOtherCode = \`console.log('test');\`;
export const defaultCode = \`should be removed\`;
`;


  describe('when processing valid preview files', () => {
    it('should generate file with defaultCode export', async () => {
      // Arrange
      createMockPreviewFile('alert', mockAngularComponent);

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      expectGeneratedFileExists('alert');
      
      const content = getGeneratedFileContent('alert');
      expect(content).toContain('export const defaultCode = `');
      expect(content).toContain('AlertPreviewComponent');
      expect(content).toContain('spartan-alert-preview'); // from selector
    });

    it('should include proper header comments', async () => {
      // Arrange
      createMockPreviewFile('button', mockAngularComponent);

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      const content = getGeneratedFileContent('button');
      expect(content).toContain('// eslint-disable -- auto-generated');
      expect(content).toContain('// prettier-ignore -- auto-generated');
      expect(content).toContain('DO NOT EDIT THIS FILE!!');
    });

    it('should handle multiple example files', async () => {
      // Arrange
      createMockPreviewFile('toggle', mockAngularComponent);
      createMockExampleFile('toggle', 'disabled', mockAngularComponent.replace('AlertPreviewComponent', 'DisabledComponent'));

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      const content = getGeneratedFileContent('toggle');
      expect(content).toContain('export const defaultCode = `');
      expect(content).toContain('export const disabled = `'); // from filename
      expect(content).toContain('AlertPreviewComponent'); // from preview
      expect(content).toContain('DisabledComponent'); // from example
    });
  });

  describe('when filtering unwanted exports', () => {
    it('should remove export const with Code suffix', async () => {
      // Arrange
      createMockPreviewFile('test', mockComponentWithUnwantedExports);

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      const content = getGeneratedFileContent('test');
      expect(content).toContain('TestComponent'); // Keep the actual component
      expect(content).not.toContain('defaultSkeleton'); // Remove skeleton export
      expect(content).not.toContain('defaultImports'); // Remove imports export
      expect(content).not.toContain('someOtherCode'); // Remove other code export
      expect(content).not.toContain('should be removed'); // Remove defaultCode content
    });

    it('should preserve import statements and decorators', async () => {
      // Arrange
      const componentWithImports = `
import { Component } from '@angular/core';
import { SomeService } from './service';

@Component({
  selector: 'test-comp',
  template: '<div>Test</div>'
})
export class TestComp {}

export const unwantedCode = \`remove me\`;
`;
      createMockPreviewFile('imports', componentWithImports);

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      const content = getGeneratedFileContent('imports');
      expect(content).toContain("import { Component } from '@angular/core'");
      expect(content).toContain("import { SomeService } from './service'");
      expect(content).toContain('@Component({');
      expect(content).not.toContain('unwantedCode');
      expect(content).not.toContain('remove me');
    });
  });

  describe('when handling edge cases', () => {
    it('should handle missing components directory', async () => {
      // Arrange - no components directory created

      // Act & Assert - should not throw
      await expect(extractPrimitiveCodeGenerator(tree)).resolves.not.toThrow();
      
      // No files should be generated
      expectGeneratedFileNotExists('alert');
    });

    it('should skip primitive with no preview files', async () => {
      // Arrange - create directory but no preview file
      const primitiveDir = joinPathFragments(componentsBasePath, 'empty');
      tree.write(joinPathFragments(primitiveDir, 'some-other-file.ts'), '// not a preview');

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      expectGeneratedFileNotExists('empty');
    });

    it('should handle file read errors gracefully', async () => {
      // Arrange - create a directory that looks like a file
      const badPath = joinPathFragments(componentsBasePath, 'bad', 'bad.preview.ts');
      // Note: not creating the actual file, just the directory structure

      // Act & Assert - should not throw
      await expect(extractPrimitiveCodeGenerator(tree)).resolves.not.toThrow();
      expectGeneratedFileNotExists('bad');
    });

    it('should continue processing other primitives if one fails', async () => {
      // Arrange
      createMockPreviewFile('good', mockAngularComponent);
      // Create a primitive directory but don't create proper files for 'bad'
      tree.write(joinPathFragments(componentsBasePath, 'bad', 'not-preview.ts'), 'invalid');

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      expectGeneratedFileExists('good'); // Should still process the good one
      expectGeneratedFileNotExists('bad'); // Should skip the bad one
    });
  });

  describe('when handling primitive name variations', () => {
    it('should handle primitive names with parentheses', async () => {
      // Arrange
      createMockPreviewFile('(special)', mockAngularComponent);

      // Act
      await extractPrimitiveCodeGenerator(tree);

      // Assert
      // Should create file with cleaned name (no parentheses)
      const filePath = joinPathFragments(componentsBasePath, '(special)', 'special.generated.ts');
      expect(tree.exists(filePath)).toBe(true);
    });
  });
});
```

elite-benni replied:
I think these are pretty good

goetzrobin commented on pnpm-lock.yaml:
I wonder why this changed. Shouldn't this stay the same?

elite-benni commented on libs/tools/src/generators/generate-primitive-snippets/generator.ts:
yes you are right, the code grew a little bit too unreadability. 
yes it nearly still works the same and is more readable. 
Altough it introduced bugs and unwanted additions. :)

elite-benni commented:
Refactored generator and tests with your generated code and modified it a little bit. 
Thanks for pointing that out! 
After updating all primitives, i lost the focus of checking the generator again. :/

goetzrobin commented:
@elite-benni looks great now!
